src/core.c/Rakudo/Internals.pm6

```diff
@@ -1383,18 +1383,15 @@ my class Rakudo::Internals {
         method pull-one() {
             nqp::while(
               nqp::chars(my str $entry = self!next),
-              nqp::if(
-                nqp::stat(
-                  (my str $path = nqp::concat($!abspath,$entry)),
-                  nqp::const::STAT_EXISTS
-                ),
+              nqp::stmts(
+                (my str $path = nqp::concat($!abspath,$entry)),
                 nqp::if(
-                  nqp::stat($path,nqp::const::STAT_ISREG)
-                    && $!file.ACCEPTS($entry),
+                  $!file.ACCEPTS($entry) &&
+                    (try nqp::stat($path,nqp::const::STAT_ISREG)),
```

Comment:
Won't this `try` slow things down significantly?

Reply:
not as much as the additional `stat` calls apparently - see the commit message that mentions a 10% speed up

Reply:
The other thing is that it is technically more correct with the `try` because there is a race condition between the old stat on existence and the stat to check if its a file or directory.

Reply:
True.

Wouldn't it make more sense to have a single `try` wrapping the whole thing?

Reply:
No. The desired behavior is not to abort entirely once an exception occurs. If you look at how it is implemented in this PR we would expect the `nqp::stat($path,nqp::const::STAT_ISREG)` to throw an exception yet we still want to continue to the next stat call to test if its a directory.

Reply:
Actually maybe it makes sense. I was confused in my previous statement because an exception would not occur when the path exists but it is a directory.